### ISSUE TITLE
Texturing mesh from images?

### ISSUE THREAD

**Issue by dzak83** [BrunoLevy/geogram#59]
Hi,
Thanks for working on this and making it available for us. I've been trying to get Poisson reconstruction for some time and I couldn't be happier to see it's working in your implementations! 

I have a question about texturing from [here](https://github.com/BrunoLevy/geogram/wiki/Pipeline). I am trying to texture the surface from the photos. I have a depth map form the Lidar scan and corresponding RGB images. I can use your pipeline to convert it from depth map to point cloud to mesh and generate mesh atlas. Is geogram supporting texture mapping from the images?

**Comment by BrunoLevy** [comment]
Hi,
There are several ways of doing that. 
 One possitibility:
  - You have your reconstructed mesh (let us call it $S$)
  - You have your point data $(x_i,y_i,z_i)$ and the associated colors $(r_i,g_i,b_i)$, you put them in a KDtree
  - You create a texture atlas for $S$
  - For each texel $(u,v)$ of $S$, you find the corresponding point $x(u,v) y(u,v) z(u,v)$ and you find the nearest $(x_i,y_i,z_i)$ of the KDtree. You use $(r_i,g_i,b_i)$ for the color of the texel. 

This is more or less what
[`bake_mesh_facet_normals_indirect()`](https://brunolevy.github.io/geogram/namespaceGEO.html#a3ef553be7c6a9b77ef91832852150ec1) and 
[`bake_mesh_points_attribute_indirect()`](https://brunolevy.github.io/geogram/namespaceGEO.html#a6342cf376576fdef825f687e90f1701f) do (you can take a look at their source to get inspiration). 

If you are stuck tell me, and I'll write a `bake_mesh_colors_indirect()` function.

**Comment by BrunoLevy** [comment]
TODO: add paragraph in documentation.

**Comment by dzak83** [comment]
Thank you! I will go deeper into that part of code. Do I understand it correctly that `MeshGrobSurfaceCommands::bake_colors` is baking just colors at the verts but void `MeshGrobSurfaceCommands::bake_texture` is mapping the whole texture? 

I also found this project: https://github.com/nmoehrle/mvs-texturing ([PDF](https://download.hrz.tu-darmstadt.de/pub/FB20/GCC/paper/Waechter-2014-LTB.pdf)) which is mapping textures on 3D objects. They are doing a lot of post applying improvements for texture differences. Does Geogram doing additional post processing? 

Thanks again for replying so quickly!

**Comment by BrunoLevy** [comment]
For atlas generation from multi-view image, yes with my group we developed an efficient solution, something that shares some similarities with the link you sent, but with several original algorithms that we developed, to hide the seams, to fill the zones that were not captured, to improve the contrast, to correct small registration errors...

This solution is part of Vorpaline, a proprietary software library developed on top of geogram, marketed by the [Tessael](https://www.tessael.com/) company. See also this [press release](https://www.inria.fr/sites/default/files/2020-10/20201002_CP_Tessael_ENrelu.pdf). Do not hesitate to contact them if you are interested.
